gadopt/gplates.py

```diff
@@ -0,0 +1,449 @@
+            self.function_space(),
+            self,
+            sub_domain=self.top_boundary_marker)
+        # coordinates of surface points
+        self.boundary_coords = fd.Function(
+            self.function_space(),
+            name="coordinates").interpolate(
+                fd.SpatialCoordinate(
+                    extract_unique_domain(self))).dat.data_ro_with_halos[self.dbc.nodes]
+        self.boundary_coords /= np.linalg.norm(self.boundary_coords, axis=1)[:, np.newaxis]
+
+        # Store the GPlates connector
+        self.gplates_connector = gplates_connector
+
+
+class pyGplatesConnector(object):
+    # Non-dimensionalisation constants
+    # Factor to non-dimensionalise gplates velocities: d/kappa
+    velocity_non_dim_factor = 2890e3/1.0e-6
+    # Factor to dimensionalise model time: d^2/kappa
+    time_dim_factor = 2890e3**2/1.0e-6
+    # 1 yr in seconds
+    yrs2sec = 365*24*60*60
+    # 1 Myr in seconds
+    myrs2sec = 1e6*yrs2sec
+
+    def __init__(self, rotation_filenames, topology_filenames, geologic_zero, delta_time=1., scaling_factor=1.0, nseeds=1000, nneighbours=3):
+        """
+        An interface to pygplates, used for updating top Dirichlet boundary conditions
+        using plate tectonic reconstructions.
+
+        This class provides functionality to assign plate velocities at different geological
+        times to the boundary conditions specified with dbc. Due to potential challenges in
+        identifying the plate id for a given point with pygplates, especially in high-resolution
+        simulations, this interface employs a method of calculating velocities at a number
+        of equidistant points on a sphere. It then interpolates these velocities for points
+        assigned a plate id. A warning is raised for any point not assigned a plate id.
+
+        Attributes:
+            rotation_filenames (str or list of str): Collection of rotation file names for pygplates.
+            topology_filenames (str or list of str): Collection of topology file names for pygplates.
+            geologic_zero (float): The oldest time present in the plate reconstruction model.
+            delta_time (float, optional): The time window range outside which plate velocities are updated. Defaults to 1.0.
+            scaling_factor (float, optional): Scaling factor for surface velocities. Defaults to 1.0.
+            nseeds (int, optional): Number of seed points used in the Fibonacci sphere generation. Defaults to 1000.
+            nneighbours (int, optional): Number of neighboring points to consider in velocity calculations. Defaults to 3.
+
+        Methods:
+            assign_plate_velocities(model_time):
+                Assigns plate velocities based on the specified model time.
+            ndtime2geotime(ndtime):
+                Converts non-dimensionalised time to geologic time.
+            geotime2ndtime(geotime):
+                Converts geologic time to non-dimensionalised time.
```

Comment:
These don't need to be here (at the class level) if there are docstrings for the methods themselves.